[CS-2750] Service degradation notice

oidoug opened the pull request:
<!-- This is a pull request. Please make sure any applicable bullet points have been covered. -->

### Description

New `Notice` Component that will be (preferably) rendered inside the lower area of the navigation bar. `Notice` is reusable and will work similarly to a Toast component, but it's fixed and can be used as a button or link.

New `ServiceStatusNotice` uses `Notice` and will handle fetching the network status information to be displayed.

- [x] Completes #(CS-2750)

### Checklist

- [x] All UI changes have been tested on a small device

### Screenshots

<!-- Screenshots or animated GIFs included here -->

https://user-images.githubusercontent.com/129619/148952344-2ad36648-0f34-455a-b681-cb620f29ae02.mp4

<!-- Use this tag to format the screenshot size

<img width="300" alt="Screenshot" src="URL_GOES_HERE"> -->

Commit:
Updated launch script for Android:
 - 'yarn android' does not uninstall the app anymore,
 - 'yarn android:clean' uninstalls and runs clean.

Commit:
New Notice base component for Notice "banner" alerts.
WIP: ServiceStatusNotice for pulling status and rendering Notice.

lukemelia commented:
Looking good so far!

danibonilha commented on cardstack/src/components/Notice/Notice.tsx:
```diff
@@ -0,0 +1,73 @@
+  icon,
+}: NoticeProps) => {
+  return (
+    <AnimatedPressable onPress={onPress} testID="notice-pressable">
```
I think we can use a simple `Touchable` here, bc it's more of a `toast` than a button itself, or even a touchable without feedback ? 🤔

oidoug replied:
Yes, I was considering the same. Maybe a Touchable will feel more in line with all the other elements in the wallet screen.

danibonilha commented on package.json:
```diff
@@ -4,7 +4,8 @@
     "setup": "yarn install && yarn postinstall",
     "build:ios": "node node_modules/react-native/local-cli/cli.js bundle --entry-file='index.js' --bundle-output='./ios/main.jsbundle' --dev=false --platform='ios'",
-    "android": "adb uninstall com.cardstack.cardpay || true && react-native run-android",
+    "android": "react-native run-android",
+    "android:clean": "adb uninstall com.cardstack.cardpay || yarn android",
```
thanks 👏

Commit:
Created service status api with RTKQuery,
along with IncidentType and it's configurations.
Finished ServiceStatusNotice comp to handle incident data.

Commit:
Merge branch 'develop' into service-degradation-notice-in-mobile-app-cs-2750

Commit:
Added service status refetch on pull-to-refresh from AssetsList

oidoug commented:
https://user-images.githubusercontent.com/129619/149335882-4888cc08-c07c-4f25-844e-136ee57a2c5f.mp4

We might need to fake some incidents for you to test this properly.
During development, I used a hard-coded fake json server for it.

danibonilha commented on cardstack/src/components/Notice/Notice.tsx:
```diff
@@ -0,0 +1,99 @@
+import { NoticeType } from '@cardstack/types';
+import { ColorTypes } from '@cardstack/theme';
+
+if (
+  Platform.OS === 'android' &&
+  UIManager.setLayoutAnimationEnabledExperimental
+) {
+  UIManager.setLayoutAnimationEnabledExperimental(true);
+}
+
```
we have this set globally on `App.js`

oidoug replied:
Ok, I also took the chance to improve the animation a bit.

danibonilha commented on cardstack/src/components/Notice/ServiceStatusNotice.tsx:
```diff
@@ -0,0 +1,20 @@
+      type="warning"
+      onPress={() => {
+        Linking.openURL(`https://status.cardstack.com`);
+      }}
```
we can wrap this on useCallback to avoid recreating on new render

danibonilha commented on cardstack/src/services/service-status-api.ts:
```diff
@@ -0,0 +1,47 @@
+
+  return (
+    major.sort((a: IncidentType, b: IncidentType) => {
+      return +new Date(b.started_at) - +new Date(a.started_at);
```
nit: maybe this could be a helper function since we are using twice, we also have lodash in the project,  if you're familiar with it

oidoug replied:
I moved the incident filter to service/utils refactored it with lodash to work with any incident type (not only critical or major), and added some tests.

danibonilha commented on cardstack/src/services/service-status-api.ts:
```diff
@@ -0,0 +1,47 @@
+  baseQuery: fetchBaseQuery({ baseUrl: STATUS_PAGE_URL }),
+  tagTypes: [...Object.values(ServiceStatusTags)],
+  endpoints: builder => ({
+    getServiceStatus: builder.query({
```
we can type the response right after `build.query<SomeResponseType, void>`  than `transformResponse` won't need to be any
Ps. this should probably remove the necessity to pass null as an argument for the query

danibonilha commented on cardstack/src/components/AssetList/AssetList.tsx:
```diff
@@ -51,6 +52,7 @@ const onScrollToIndexFailed = () => {
 export const AssetList = (props: AssetListProps) => {
   const sectionListRef = createRef<SectionList>();
   const refresh = useRefreshAccountData();
+  const { refetch: refetchServiceStatus } = useGetServiceStatusQuery(null);
```
Why do we need to pass null? maybe we are missing to type something

danibonilha reviewed (commented):
Props on using RTK 👏 , it's really looking good, just a couple of suggestions, it would be nice to add on this PR the frequency to refetch the status api also, if there are any queries we should invalidate the status cache to be refetched, or if refetching every hour or so is enough, or just once the app first starts.

danibonilha commented on cardstack/src/components/Notice/Notice.tsx:
```diff
@@ -0,0 +1,99 @@
+};
+
+const iconColorByType: {
+  [key in NoticeType]: ColorTypes;
```
nit: this could be extracted to it's own type, so we can reuse on the others styles
``` ts
type NoticeVariant = { [key in NoticeType]: ColorTypes } 
```
or
``` ts
type NoticeVariant = Record<NoticeType, ColorTypes>
```
or we could even use one noticeConfig with all the variants in it

``` tsx
interface NoticeStyle {
  iconColor: ColorTypes;
  backgroundColor: ColorTypes;
  textColor: ColorTypes;
}


const noticeConfig: Record<NoticeType, NoticeStyle> = {
  info: {
    iconColor: 'appleBlue',
    backgroundColor: 'backgroundGray',
    textColor: 'black',
  },
  ...
};

<>
  backgroundColor={noticeConfig[type].backgroundColor}
</>
```

oidoug replied:
Loved that last suggestion, so much cleaner :) Thanks for the quality review, Dani!

Commit:
Fixes service-status-api rtkquery return and param types;
Improves animation on Notice visibility;
Adds tests for filter-incident utils;
Moves query-promise-wrapper from utils index to own file.

danibonilha commented on cardstack/src/components/Notice/Notice.tsx:
```diff
@@ -56,12 +43,14 @@ export const Notice = ({
+  useLayoutEffect(() => {
+    LayoutAnimation.configureNext(
+      LayoutAnimation.create(200, 'easeInEaseOut', 'opacity')
+    );
```
we have this as a helper function => `layoutEasingAnimation`

danibonilha reviewed (approved):
LGTM, just one small "fix", congrats on your first feature 🚀

Commit:
Using helper function for layout animation.

lukemelia commented:
Way to go @douglaslondrina! 👏 👏 👏  🥇
